More loop variants.

DrTodd13 opened the pull request:
Add teams loop and bare loop directives.  Add tests for those but the teams loop one is skipped for now due to need for backend support.  Fix some bugs related to having separate directives instead of combined.

DrTodd13 committed:
Add teams loop directive.

DrTodd13 committed:
1) Add additional vars, that were previously missing, created during loop and collapse processing to the enclosing openmp region.
2) Fix bug that checks for distribute being tightly nested within teams.
3) Add bare loop directive.
4) Don't add QUAL.OMP.THREAD_LIMIT on non-team parallel directives.
5) Add test for separate target, teams, and loop.

ggeorgakoudis commented:
Check github testing, it's failing in python in running on CPU (which should not fail):
```
    tags_for_enclosing = [cmp_var.name, omp_lb_var.name, omp_start_var.name, omp_iv_var.name, types_mod_var.name, int64_var.name, itercount_var.name, omp_ub_var.name, const1_var.name, const1_latch_var.name, get_itercount_var.name] + [x.name for x in iterspace_vars]
UnboundLocalError: Failed in nopython mode pipeline (step: Handle with contexts)
local variable 'iterspace_vars' referenced before assignment
```

DrTodd13 committed:
Move collapse related var outside collapse region so that it is available for tags_for_enclosing when there is no collapse clause.

ggeorgakoudis commented on numba/openmp.py:
```diff
@@ -532,7 +532,7 @@ def replace_vars_inner(self, var_dict):
     def add_to_usedef_set(self, use_set, def_set, start):
         assert start==True or start==False
-        if config.DEBUG_OPENMP >= 1:
+        if config.DEBUG_OPENMP >= 3:
```
Is "3" intentional?

DrTodd13 replied:
Yes.

ggeorgakoudis commented on numba/tests/test_openmp.py:
```diff
@@ -3881,6 +3881,44 @@ def test_impl():
         r = test_impl()
         np.testing.assert_array_equal(r, np.arange(10) * 10)
 
+    @unittest.skip("Fix backend")
```
What is the error?

DrTodd13 replied:
DIR.OMP.TEAMS.DISTRIBUTE.PARALLEL.LOOP  directive is not supported
Going to add a comment after the skip to say what the reason we are skipping is.

ggeorgakoudis commented on numba/tests/test_openmp.py:
```diff
@@ -3881,6 +3881,44 @@ def test_impl():
         np.testing.assert_array_equal(r, np.arange(10) * 10)
 
+    @unittest.skip("Fix backend")
+    def target_separate_teams_loop(self, device):
```
Use the existing convention for naming, when directives are nested, we write it as `<directive1>_nest_<directive2>`, when they are combined we write it as `<directive1>_<directive>2`. So, in this case, `target_nest_teams_loop`

DrTodd13 replied:
Done.

ggeorgakoudis commented on numba/tests/test_openmp.py:
```diff
@@ -3881,6 +3881,44 @@ def test_impl():
+        c = test_impl(n)
+        np.testing.assert_array_equal(c, np.full((n,n), 2))
+
+    def target_teams_loop_separate(self, device):
```
Naming, `target_nest_teams_nest_loop`

DrTodd13 replied:
Done.

DrTodd13 committed:
Adding way to skip test only for device=1.

DrTodd13 committed:
1) Remove some old commented code.
2) Add functions to process schedule_clause and dist_schedule_clause from the grammar.
3) Change loop directive to look at enclosing region and generate distribute parallel for if in a teams region and parallel for otherwise.
4) Add kind of failure to all CPU-target-backend skipped cases.
5) Add test for loop directive on CPU side.

DrTodd13 committed:
If loop directive under target without explicit teams than an implicit teams.

DrTodd13 committed:
Fix map problem where default tofrom is used.  Improve loop a bit such that if inside distribute then just parallel loop is used.

DrTodd13 committed:
Add comment.

DrTodd13 committed:
Add support for combined target loop directive.

ggeorgakoudis commented on numba/tests/test_openmp.py:
Remove `skipIf` for tests on `device(1)`. Fixed to be runnable in #36. Make only additive changes to `test_openmp.py`

DrTodd13 replied:
Done.

DrTodd13 committed:
Remove the skipping of target cpu device emulation for certain tests.

ggeorgakoudis commented:
@DrTodd13 Checks pass! There's some conflicts to resolve. LGTM.
